File: server/src/routes/knowledge.js
// server/src/routes/knowledge.js
// 知识库路由：文档管理（上传/列表/删除）+ RAG 问答（流式）
import express from 'express'
import multer from 'multer'
import path from 'path'
import { ingestDocument, getDocRegistry, deleteDocument } from '../services/rag/ingest.js'
import { ragQueryStream } from '../services/rag/query.js'
import { rateLimiter } from '../middleware/index.js'
import { sendSseError } from '../utils/errors.js'
import { logger } from '../utils/logger.js'

export const knowledgeRouter = express.Router()

// ── multer 文件上传配置 ────────────────────────────────────────
// memoryStorage：文件先存内存，再由业务代码决定怎么处理
// diskStorage：直接存磁盘（大文件推荐）
const upload = multer({
  storage: multer.diskStorage({
    destination: (req, file, cb) => cb(null, './uploads/'),
    filename:    (req, file, cb) => {
      const ext  = path.extname(file.originalname)
      const name = `${Date.now()}_${Math.random().toString(36).slice(2, 6)}${ext}`
      cb(null, name)
    },
  }),
  limits: {
    fileSize: 10 * 1024 * 1024,  // 最大 10MB
  },
  fileFilter: (req, file, cb) => {
    // 只允许这几种格式
    const allowed = ['.txt', '.md', '.pdf']
    const ext = path.extname(file.originalname).toLowerCase()
    if (allowed.includes(ext)) {
      cb(null, true)
    } else {
      cb(new Error(`不支持的文件格式 ${ext}，只支持 ${allowed.join(', ')}`))
    }
  },
})

// ── POST /api/knowledge/documents ─────────────────────────────
// 上传文档并入库
// 支持两种方式：1) 上传文件  2) 直接传文本内容
knowledgeRouter.post('/documents', rateLimiter, async (req, res) => {
  // 先尝试文件上传
  upload.single('file')(req, res, async (uploadErr) => {
    try {
      let docMeta

      if (req.file) {
        // 方式1：上传文件
        docMeta = await ingestDocument({
          filePath: req.file.path,
          fileName: req.file.originalname,
          title:    req.body.title || req.file.originalname.replace(/\.[^.]+$/, ''),
          category: req.body.category || '通用',
          mimeType: req.file.mimetype,
        })
      } else if (req.body.content) {
        // 方式2：直接传文本（前端粘贴内容）
        // 先写到临时文件，再走统一入库流程
        const fs = await import('fs/promises')
        const tmpPath = `./uploads/tmp_${Date.now()}.txt`
        await fs.writeFile(tmpPath, req.body.content, 'utf-8')

        docMeta = await ingestDocument({
          filePath: tmpPath,
          fileName: (req.body.title || '文本内容') + '.txt',
          title:    req.body.title || '未命名文档',
          category: req.body.category || '通用',
          mimeType: 'text/plain',
        })
      } else {
        return res.status(400).json({ error: { message: '请上传文件或提供文本内容' } })
      }

      res.json({ success: true, document: docMeta })
    } catch (err) {
      logger.error('knowledge: ingest error', { error: err.message })
      res.status(500).json({ error: { message: err.message || '文档处理失败' } })
    }
  })
})

// ── GET /api/knowledge/documents ──────────────────────────────
// 获取所有已入库的文档
knowledgeRouter.get('/documents', (req, res) => {
  const docs = getDocRegistry()
  const { category } = req.query

  const filtered = category
    ? docs.filter(d => d.category === category)
    : docs

  res.json({ documents: filtered })
})

// ── DELETE /api/knowledge/documents/:docId ─────────────────────
// 删除文档（从向量库和注册表中删除）
knowledgeRouter.delete('/documents/:docId', async (req, res) => {
  try {
    await deleteDocument(req.params.docId)
    res.json({ success: true })
  } catch (err) {
    res.status(404).json({ error: { message: err.message } })
  }
})

// ── POST /api/knowledge/query/stream ──────────────────────────
// RAG 问答（流式）：先推送来源，再流式推送回答
knowledgeRouter.post('/query/stream', rateLimiter, async (req, res) => {
  const { question, category } = req.body

  if (!question?.trim()) {
    return res.status(400).json({ error: { message: '问题不能为空' } })
  }

  res.setHeader('Content-Type', 'text/event-stream')
  res.setHeader('Cache-Control', 'no-cache')
  res.setHeader('Connection', 'keep-alive')
  res.setHeader('X-Accel-Buffering', 'no')

  const send = (event, data) => {
    if (!res.writableEnded) {
      res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`)
    }
  }

  try {
    send('status', { message: '正在检索相关文档...' })

    const { sources, streamAnswer } = await ragQueryStream(question, { category })

    // 先推送检索到的来源（让用户看到在参考哪些文档）
    send('sources', { sources })

    if (!sources.length) {
      send('token', { token: '知识库中未找到相关内容，请尝试上传相关文档后再提问。' })
      send('done', {})
      return res.end()
    }

    send('status', { message: '正在生成回答...' })

    // 流式推送回答
    for await (const token of streamAnswer()) {
      send('token', { token })
    }

    send('done', {})
    logger.info('knowledge: query done', {
      question: question.slice(0, 40),
      sources:  sources.length,
    })
  } catch (err) {
    logger.error('knowledge: query error', { error: err.message })
    sendSseError(res, err)
  } finally {
    if (!res.writableEnded) res.end()
  }
})

// ── GET /api/knowledge/categories ─────────────────────────────
// 获取所有分类（前端筛选用）
knowledgeRouter.get('/categories', (req, res) => {
  const docs = getDocRegistry()
  const categories = [...new Set(docs.map(d => d.category))]
  res.json({
    categories: [
      { value: '', label: '全部文档' },
      ...categories.map(c => ({ value: c, label: c })),
    ],
  })
})

File: server-py/app/routes/knowledge.py
# server-py/app/routes/knowledge.py
# 知识库路由：文档管理（上传/列表/删除）+ RAG 问答（流式）
import os
import random
import time

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile

from app.middleware import rate_limiter
from app.services.rag.ingest import delete_document, get_doc_registry, ingest_document
from app.services.rag.query import rag_query_stream
from app.utils.logger import logger
from app.utils.sse import sse_stream

router = APIRouter()
# "./uploads"是一个相对路径字符串，表示「当前工作目录下的 uploads 文件夹」：，而不是代码文件所在位置。所以：
# 如果你在 server-py/ 目录下启动服务（如 uvicorn app.main:app），上传目录会是 server-py/uploads
# 如果你从项目根目录 ai_project_node/ 启动，则会是 ai_project_node/uploads
UPLOAD_DIR = "./uploads"
os.makedirs(UPLOAD_DIR, exist_ok=True)

ALLOWED_EXTENSIONS = {".txt", ".md", ".pdf"}
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB


def _gen_filename(original_name: str) -> str:
    ext = os.path.splitext(original_name)[1]
    rand = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=4))
    return f"{int(time.time() * 1000)}_{rand}{ext}"


@router.post("/documents", summary="上传知识库文档", dependencies=[Depends(rate_limiter)])
async def upload_document(
    file: UploadFile | None = File(default=None),
    content: str | None = Form(default=None),
    title: str | None = Form(default=None),
    category: str | None = Form(default="通用"),
):
    try:
        if file:
            ext = os.path.splitext(file.filename or "")[1].lower()
            if ext not in ALLOWED_EXTENSIONS:
                raise HTTPException(status_code=400, detail={
                    "error": {"message": f"不支持的文件格式 {ext}，只支持 {', '.join(ALLOWED_EXTENSIONS)}"}})

            data = await file.read()
            if len(data) > MAX_FILE_SIZE:
                raise HTTPException(status_code=400, detail={"error": {"message": "文件超过 10MB 限制"}})

            saved_name = _gen_filename(file.filename or "upload")
            saved_path = os.path.join(UPLOAD_DIR, saved_name)
            with open(saved_path, "wb") as f:
                f.write(data)

            doc_title = title or os.path.splitext(file.filename or "")[0]
            doc_meta = await ingest_document(
                file_path=saved_path, file_name=file.filename or saved_name,
                title=doc_title, category=category or "通用",
            )
        elif content:
            tmp_path = os.path.join(UPLOAD_DIR, f"tmp_{int(time.time() * 1000)}.txt")
            with open(tmp_path, "w", encoding="utf-8") as f:
                f.write(content)

            doc_meta = await ingest_document(
                file_path=tmp_path, file_name=(title or "文本内容") + ".txt",
                title=title or "未命名文档", category=category or "通用",
            )
        else:
            raise HTTPException(status_code=400, detail={"error": {"message": "请上传文件或提供文本内容"}})

        return {"success": True, "document": doc_meta}
    except HTTPException:
        raise
    except Exception as err:
        logger.error("knowledge: ingest error", {"error": str(err)})
        raise HTTPException(status_code=500, detail={"error": {"message": str(err) or "文档处理失败"}})


@router.get("/documents")
async def list_documents(category: str | None = None):
    docs = get_doc_registry()
    if category:
        docs = [d for d in docs if d["category"] == category]
    return {"documents": docs}


@router.delete("/documents/{doc_id}")
async def remove_document(doc_id: str):
    try:
        await delete_document(doc_id)
        return {"success": True}
    except ValueError as err:
        raise HTTPException(status_code=404, detail={"error": {"message": str(err)}})


@router.post("/query/stream", dependencies=[Depends(rate_limiter)])
async def query_stream(body: dict):
    question = (body.get("question") or "").strip()
    category = body.get("category")

    if not question:
        raise HTTPException(status_code=400, detail={"error": {"message": "问题不能为空"}})

    async def generator():
        yield "status", {"message": "正在检索相关文档..."}

        sources, stream_answer = await rag_query_stream(question, category)

        yield "sources", {"sources": sources}

        if not sources:
            yield "token", {"token": "知识库中未找到相关内容，请尝试上传相关文档后再提问。"}
            yield "done", {}
            return

        yield "status", {"message": "正在生成回答..."}

        async for token in stream_answer():
            yield "token", {"token": token}

        yield "done", {}
        logger.info("knowledge: query done", {"question": question[:40], "sources": len(sources)})

    return sse_stream(generator)


@router.get("/categories")
async def categories():
    docs = get_doc_registry()
    seen = []
    for d in docs:
        if d["category"] not in seen:
            seen.append(d["category"])
    return {"categories": [{"value": "", "label": "全部文档"}, *[{"value": c, "label": c} for c in seen]]}
